docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+
+## Proposal
+
+Extending the analogy of [node-problem-detector](#node-problem-detector) to the larger granularity of the Kubernetes cluster, the [`ClusterCondition`s](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L147) in the[`status`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L66) of the [`Cluster`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L30) custom resource (as defined in the [cluster-registry](#cluster-registry)) can be used as the standard way to publish and consume health conditions that have cluster-wide implication.
+
+I.e.
+- A single `Cluster` resource instance is maintained for each cluster in the shoot control-plane namespace of the seed cluster.
+- Many different monitoring components (either dedicated to each shoot cluster or deployed as a common component in the seed clusters) then update the `status.conditions` on the `Cluster` resource instance(s).
+- Many different remedy systems watch the `status.conditions` (either dedicated to each shoot cluster or deployed as a common component in the seed cluster) on the `Cluster` resource instance(s) for conditions of (one or more) specific condition types and take remedial action(s). 
+
+With this approach, the monitoring components that update the health conditions can be decoupled from the remedy system components that react to such health conditions and take some remedial action.
+To be clear,
+- One monitoring component can be responsible for updating more than one health condition type.
+- More than one remedy system component can take possible remedial action for any given health condition.
+
+### Condition Types of immediate relevance
+
+This is just a sample set of condition types that might be of immediate relevance. It is not an exhaustive list.
+The suggested possible actions for these condition types are also just samples and not an exhaustive list.
+
+#### KubeAPIServerUnreachableExternally
+
+The monitoring (probing) part of the current `probe` [sub-command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/probe.go) of the [dependency-watchdog](ttps://github.com/gardener/dependency-watchdog) can be re-factored into a separate component that updates the conditions with the `KubeAPIServerUnreachableExternally` condition type.
```

Comment:
> API probes are supported by the shoot care controller and it updates the shoot status with the result.
The monitoring stack is also watching them and can fire alerts, so what will be the added value here, if one more component is watching the API server status?

I am not proposing a new component here. This component already [exists](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/probe-deployment.yaml). It monitors the shoot `kube-apiserver` endpoint both via the `ClusterIP` as well as the `LoadBalancer`. Shoot care controller checks only one of these.

> I do not see what general cloud agnostic remedy action can be implemented here, except try to re-create kube-apiserver service, but should do nothing in case quota/rate limits are reached.

It [acts](https://github.com/gardener/dependency-watchdog/blob/master/pkg/scaler/prober.go#L230) by scaling down the `kube-controller-manager`. I am not a fan of this but that seems like the only option. Otherwise, the `nodelifecycle` controller will [mark](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L860) all pods in all the unhealthy nodes as `NotReady`. If the `LoadBalancer` is down, every `node` in the cluster would be marked ready as soon as the `--node-monitor-grace-period` elapses and hence all pods in the cluster will be marked `NotReady`. This is not desirable. Control-plane issues should not unnecessarily disrupt existing workload (as long as they are not dependent on the control-plane themselves).

The only alternative would have been to patch the `kube-controller-manager` `Deployment` to disable the `nodelifecycle` controller. The correct long-term solution is to file a PR on the `nodelifecycle` controller, of course.